pkg/3scale/amp/operator/system_reconciler.go

```diff
@@ -182,6 +143,134 @@ func (r *SystemReconciler) Reconcile() (reconcile.Result, error) {
+		err = helper.DeleteJob(component.SystemAppPreHookJobName, r.apiManager.GetNamespace(), r.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	generationChanged, err = helper.HasAppGenerationChanged(component.SystemAppPostHookJobName, component.SystemAppDeploymentName, r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if generationChanged {
+		err = helper.DeleteJob(component.SystemAppPostHookJobName, r.apiManager.GetNamespace(), r.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Used to synchronize the system-app Deployment with the PreHook/PostHook Jobs
+	currentAppDeploymentGeneration, err := getSystemAppDeploymentGeneration(r.apiManager.GetNamespace(), r.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// SystemApp PreHook Job
+	preHookJob := system.AppPreHookJob(ampImages.Options.SystemImage, currentAppDeploymentGeneration)
+	err = r.ReconcileJob(preHookJob, reconcilers.CreateOnlyMutator)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Block reconciling system-app Deployment until PreHook Job has completed
+	if !helper.HasJobCompleted(preHookJob.Name, preHookJob.Namespace, r.Client()) {
+		systemComponentNotReady = true
+	}
+
+	// SystemApp Deployment
+	systemAppDeploymentMutators := []reconcilers.DMutateFn{
+		reconcilers.DeploymentAffinityMutator,
+		reconcilers.DeploymentTolerationsMutator,
+		reconcilers.DeploymentPodTemplateLabelsMutator,
+		reconcilers.DeploymentPriorityClassMutator,
+		reconcilers.DeploymentTopologySpreadConstraintsMutator,
+		reconcilers.DeploymentPodTemplateAnnotationsMutator,
+		r.systemAppDeploymentResourceMutator,
+		reconcilers.DeploymentRemoveDuplicateEnvVarMutator,
+		reconcilers.DeploymentPodContainerImageMutator,
+	}
+	if r.apiManager.Spec.System.AppSpec.Replicas != nil {
+		systemAppDeploymentMutators = append(systemAppDeploymentMutators, reconcilers.DeploymentReplicasMutator)
+	}
+	if !systemComponentNotReady {
+		err = r.ReconcileDeployment(system.AppDeployment(ampImages.Options.SystemImage), reconcilers.DeploymentMutator(systemAppDeploymentMutators...))
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Block reconciling PostHook Job unless BOTH the PreHook Job has completed and the system-app Deployment is ready and not in the process of updating
+	deployment := &k8sappsv1.Deployment{}
+	err = r.Client().Get(context.TODO(), k8sclient.ObjectKey{
+		Namespace: r.apiManager.GetNamespace(),
+		Name:      component.SystemAppDeploymentName,
+	}, deployment)
+	if err != nil && !k8serr.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
+	if k8serr.IsNotFound(err) || !helper.IsDeploymentAvailable(deployment) || helper.IsDeploymentProgressing(deployment) || !helper.HasJobCompleted(preHookJob.Name, preHookJob.Namespace, r.Client()) {
+		systemComponentNotReady = true
+	}
+
+	// SystemApp PostHook Job
+	if !systemComponentNotReady {
```

Comment:
WDYT about systemComponentReady instead? 
the !systemComponentNotReady is hard to read, but no strong opinion, might be just me...

Reply:
I agree its confusing, I just pushed changes to use `systemComponentsReady` instead.